src/mono/wasm/runtime/library_channel.js

```diff
@@ -0,0 +1,147 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+Module [ 'channel' ] = {
+    _impl: class {
+        // Index constants for the communication buffer.
+        get STATE_IDX() { return 0; }
+        get MSG_SIZE_IDX() { return 1; }
+        get COMM_LAST_IDX() { return this.MSG_SIZE_IDX; }
+
+        // Communication states.
+        get STATE_SHUTDOWN() { return -1; } // Shutdown
+        get STATE_IDLE() { return 0; }
+        get STATE_REQ() { return 1; }
+        get STATE_RESP() { return 2; }
+        get STATE_REQ_P() { return 3; } // Request has multiple parts
+        get STATE_RESP_P() { return 4; } // Response has multiple parts
+        get STATE_AWAIT() { return 5; } // Awaiting the next part
+
+        constructor(msg_char_len) {
+            this.msg_char_len = msg_char_len;
+
+            const int_bytes = 4;
+            const comm_byte_len = int_bytes * (this.COMM_LAST_IDX + 1);
+            this.comm_buf = new SharedArrayBuffer(comm_byte_len);
+
+            // JavaScript character encoding is UTF-16.
+            const char_bytes = 2;
```

Comment:
No idea if it is relevant, but I wonder whether we have tests that hash characters out of the BMP, ie., that decode to more than 2 bytes ?

Reply:
> No idea if it is relevant, but I wonder whether we have tests that hash characters out of the BMP, ie., that decode to more than 2 bytes ?

@danmoseley It shouldn't matter here as this is merely passing around a JSONified object and if it contains strings those will be encoded as UTF-16 which supports the 17 planes - it may use 2 16-bit characters instead of 1.

The second part, regarding hashing of characters outside of BMP, should exist somewhere, although I am unclear if it matters much here or if it occurs in any of the crypto hash testing at all.

Reply:
Got it - fine.